src/lib/Parser.hs

```diff
@@ -341,34 +359,41 @@ interfaceInstance = do
   (p, pos) <- withPos letPat
   ann <- annot uType
   keyWord WhereKW
-  rhs <- mkRhs ann <$> (withSrc $ URecord <$> interfaceRecordFields "=")
-  let (ann', rhs') = addImplicitImplicitArgs pos (Just ann) rhs
-  return $ ULet InstanceLet (p, ann') rhs'
+  rhsEither <- mkConstructorCall ann <$> (withSrc $ (URecord . NoExt) <$>
+      interfaceRecordFields "=")
+  case rhsEither of
+    Left err  -> fail err
+    Right rhs -> let (ann', rhs') = addImplicitImplicitArgs pos (Just ann) rhs
+                 in return $ ULet InstanceLet (p, ann') rhs'
   where
     -- Here, we are traversing the type annotation to retrieve the name of
     -- the interface and generate its corresponding constructor.
-    mkRhs (WithSrc _ (UPi _ _ lrhs))      rhs = mkRhs lrhs rhs
-    mkRhs (WithSrc _ (UApp _ (WithSrc _ (UVar v)) _)) rhs =
-      (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` rhs
-    -- We let this case through to fail during type checking.
-    mkRhs _ rhs = rhs
+    mkConstructorCall (WithSrc _ (UPi _ arr typ)) record
+      | arr `elem` [ClassArrow, ImplicitArrow] = mkConstructorCall typ record
+    -- The innermost UApp is an application of a variable that has the name of
+    -- the interface to a type. We retrieve its name and derive the name of
+    -- its automatically generated constructor, which we apply to the record
+    -- passed as a parameter.
+    mkConstructorCall (WithSrc _ (UApp _ (WithSrc _ (UVar v)) _)) record =
+      Right $ (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` record
```

Comment:
It seems like this function is responsible for both finding the constructor name and also building the call. Could you separate the two? This could just take one argument and return `Right var`, then the rest of the logic could be moved to the call site here:
https://github.com/google-research/dex-lang/blob/100184cb940899b694fb4fa5077295dc0ec81d80/src/lib/Parser.hs#L364-L367
I think that would be easier to understand.